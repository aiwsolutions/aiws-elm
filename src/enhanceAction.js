import _ from 'lodash';

const enhanceAction = (action, prefix, separator) => (...params) => {
    const actionPayload = action(...params);
    return {
        ...actionPayload,
        type: `${prefix}${separator}${actionPayload.type}`
    };
};

export default (actions, prefix, separator = '/') => {
    if (_.isArray(actions)) {
        return _.map(actions, action => enhanceAction(action, prefix, separator));
    }
    return enhanceAction(actions, prefix, separator);
};
