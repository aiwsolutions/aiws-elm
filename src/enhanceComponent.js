import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

export default (Component, reducerPrefix, actionWrapper, { storeKey = 'store', separator = '/' } = {}) =>
    class ElmComponent extends React.Component {
        static childContextTypes = {
            [storeKey]: PropTypes.object
        };
        static contextTypes = {
            [storeKey]: PropTypes.object.isRequired
        };

        getChildContext() {
            const store = this.context[storeKey];
            const actionMapping = _.isFunction(actionWrapper) ? actionWrapper :
                action => ({
                    ...action,
                    type: _.isNil(actionWrapper) ?
                        `${reducerPrefix}${separator}${action.type}` :
                        `${actionWrapper}${action.type}`
                });

            const dispatch = action => store.dispatch(actionMapping(action));

            // convert separator to dot inorder to get its proper state.
            const getState = () => _.get(store.getState(), _.replace(reducerPrefix, separator, '.'));

            return {
                [storeKey]: {
                    ...store,
                    dispatch,
                    getState
                }
            };
        }

        render() {
            return (
                <Component
                    {...this.props}
                />
            );
        }
    };
