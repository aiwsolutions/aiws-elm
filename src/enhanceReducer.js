import _ from 'lodash';

const enhanceReducer = (reducer, prefix, separator) => (state, action) => {
    const regex = new RegExp(`^(${prefix})${separator}(.+)`);
    const match = regex.exec(action.type);
    if (!match) {
        return state;
    }

    const reducerPrefix = match[1];
    const unwrappedType = match[2];
    if (!reducerPrefix || !unwrappedType) {
        return state;
    }

    const currentState = _.get(state, reducerPrefix);
    const newState = reducer(currentState, {
        ...action,
        type: unwrappedType
    });

    if (newState === currentState) {
        return state;
    }
    return {
        ...state,
        [reducerPrefix]: newState
    };
};

export default (initialState, map, separator = '/') => (state = { ...initialState }, action) => {
    const { root, ...childReducers } = map;
    const newState = _.reduce(
        childReducers,
        (acc, reducer, prefix) => enhanceReducer(reducer, prefix, separator)(acc, action),
        state
    );
    if (_.isFunction(root)) {
        return root(newState, action);
    }
    return newState;
};
