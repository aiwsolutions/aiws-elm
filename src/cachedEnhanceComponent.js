import enhanceComponent from './enhanceComponent';

let cache = {};

export function cachedEnhanceComponent(Component, reducerPrefix, actionWrapper) {
    if (cache[reducerPrefix]) {
        return cache[reducerPrefix];
    }
    const enhancedComponent = enhanceComponent(Component, reducerPrefix, actionWrapper);
    cache[reducerPrefix] = enhancedComponent;
    return enhancedComponent;
}

export function clearCache(reducerPrefix) {
    cache[reducerPrefix] = null;
}

export function resetCache() {
    cache = {};
}
