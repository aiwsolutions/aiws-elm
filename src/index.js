import enhanceReducer from './enhanceReducer';
import enhanceComponent from './enhanceComponent';
import enhanceAction from './enhanceAction';
import { cachedEnhanceComponent, resetCache, clearCache } from './cachedEnhanceComponent';

module.exports = {
    enhanceReducer,
    enhanceComponent,
    enhanceAction,
    cachedEnhanceComponent,
    resetCache,
    clearCache,
    DEFAULT_SEPARATOR: '/'
};
