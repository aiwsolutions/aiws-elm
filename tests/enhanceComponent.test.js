import React from 'react';
import { connect, Provider } from 'react-redux';
import { createStore } from 'redux';
import renderer from 'react-test-renderer';
import { enhanceReducer, enhanceComponent } from '../src';

const counterInitialState = {
    counter: 0
};

function increaseOne() {
    return {
        type: 'INCREASE_ONE'
    };
}

const counterReducer = (state = { ...counterInitialState }, action) => {
    switch (action.type) {
        case 'INCREASE_ONE':
            return {
                ...state,
                counter: state.counter + 1
            };
        default:
            return state;
    }
};

const Counter = ({ counter, id, ...others }) =>
    <div>
        {counter}
        <button
            id={`button-${id}`}
            onClick={() => others.increaseOne()}
        />
    </div>;

const mapStateToProps = state => ({
    counter: state.counter
});

const mapDispatchToProps = {
    increaseOne
};

const ConnectedCounter = connect(mapStateToProps, mapDispatchToProps)(Counter);

const racingInitialState = {
    CounterOne: counterInitialState,
    CounterTwo: counterInitialState
};

const racingReducer = enhanceReducer(racingInitialState, {
    'Counter.*': counterReducer
});

const CounterOne = enhanceComponent(ConnectedCounter, 'CounterOne');
const CounterTwo = enhanceComponent(ConnectedCounter, 'CounterTwo');

const Racing = () =>
    <div>
        <CounterOne
            id="one"
        />
        <CounterTwo
            id="two"
        />
    </div>;

test('enhanceComponent', () => {
    const store = createStore(racingReducer, {
        ...racingInitialState
    });
    const comp = renderer.create(
        <Provider store={store}>
            <Racing />
        </Provider>
    );
    expect(comp.toJSON()).toMatchSnapshot();

    // simulate button two click
    comp.root.findByProps({ id: 'button-two' }).props.onClick();
    expect(comp.toJSON()).toMatchSnapshot();
});

