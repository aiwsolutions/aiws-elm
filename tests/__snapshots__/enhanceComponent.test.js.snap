// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enhanceComponent 1`] = `
<div>
  <div>
    0
    <button
      id="button-one"
      onClick={[Function]}
    />
  </div>
  <div>
    0
    <button
      id="button-two"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`enhanceComponent 2`] = `
<div>
  <div>
    0
    <button
      id="button-one"
      onClick={[Function]}
    />
  </div>
  <div>
    1
    <button
      id="button-two"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`enhanceComponent with mutiple levels of prefix 1`] = `
<div>
  0
  <button
    id="button-counter"
    onClick={[Function]}
  />
</div>
`;

exports[`enhanceComponent with mutiple levels of prefix 2`] = `
<div>
  1
  <button
    id="button-counter"
    onClick={[Function]}
  />
</div>
`;
